ggshield/scan/repo.py

```diff
@@ -153,37 +153,33 @@ def scan_commit_range(
         )
     )
 
-    with concurrent.futures.ThreadPoolExecutor(
-        max_workers=min(CPU_COUNT, 4)
-    ) as executor:
-        future_to_process = [
-            executor.submit(
-                scan_commits_content,
-                commits,
-                client,
-                cache,
-                matches_ignore,
-                scan_context,
-                ignored_detectors,
-            )
-            for commits in commits_batch
-        ]
-
+    with click.progressbar(
+        length=len(commit_list),
+        label=format_text("Scanning Commits", STYLE["progress"]),
+        file=sys.stderr,
+    ) as bar:
         scans: List[ScanCollection] = []
-        with click.progressbar(
-            iterable=concurrent.futures.as_completed(future_to_process),
-            length=len(future_to_process),
-            label=format_text("Scanning Commits", STYLE["progress"]),
-            file=sys.stderr,
-        ) as completed_futures:
-            for future in completed_futures:
+        with ThreadPoolExecutor(max_workers=min(CPU_COUNT, 4)) as executor:
+            for future in as_completed(
+                executor.submit(
+                    scan_commits_content,
+                    commits,
+                    client,
+                    cache,
+                    matches_ignore,
+                    scan_context,
+                    ignored_detectors,
+                )
+                for commits in commits_batch
+            ):
                 scan_collection = future.result()
-                for scan in scan_collection.scans:
+                for scan in scan_collection.scans_with_results:
                     if scan.results and scan.results.errors:
                         for error in scan.results.errors:
                             # Prefix with `\n` since we are in the middle of a progress bar
                             display_error(f"\n{error.description}")
                     scans.append(scan)
+                bar.update(len(scan_collection.scans_with_results))
```

Comment:
Regarding the progress bar not moving at startup, maybe the `for` loop above takes a long time to execute? How many items are in `scan_collection.scans_with_results`?

I think you can do two changes here which may help:
- Move the `bar.update()` call before the `for` loop
- Remove the `scans.append(scan)` inside the loop with a `scans.extend(scan_collection.scans_with_results)` *outside* the loop

Reply:
Unfortunately I tested both approaches you suggest. I also had a look into using tqdm, I had the same problem.  
I will try to find another solution in the meantime.